### PR TITLE
don't trust X_DOWNLOADED_FULL flag, consider messages with missing data incomplete (WIP)

### PR DESCRIPTION
**Pull request by Valodim**
This is an experimental fix for #1243 with the following changes, the overall idea is to allow recovery from a state where part files which should be on disk are missing.

* the flag which decides whether a message is complete, and whether the "Download Complete Message" button is shown, is moved into MessageViewInfo, the relevant logic into MessageViewInfoExtractor
* if a file-backed part is missing its related file, the `FileBackedBody` won't be created (leaving the part with a null body)
* a message is considered incomplete if either it doesn't have an X_DOWNLOADED_FULL flag, *or* it is missing the body of any message part
* automatic download of (partial) message content on view is triggered only if neither X_DOWNLOADED_PARTIAL nor X_DOWNLOADED_FULL flags is set, to avoid loops if a message can't be downloaded properly
* loadMessageRemoteSynchronous will always download a message. This is currently only called exactly if the user clicks the "Download Complete Message" button, so this seems reasonable?

Feedback welcome

### CODE REVIEW

File: k9mail/src/main/java/com/fsck/k9/activity/MessageLoaderHelper.java [hunk 1]
```diff
@@ -180,7 +176,9 @@ private void onLoadMessageFromDatabaseFinished() {
 
         callback.onMessageDataLoadFinished(localMessage);
 
-        if (localMessage.isBodyMissing()) {
+        boolean messageIsIncomplete =
```
Comment (cketti) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageLoaderHelper.java]: Tiny detail: should be `messageIncomplete` so a getter would be `isMessageIncomplete()`.

File: k9mail/src/main/java/com/fsck/k9/activity/MessageLoaderHelper.java [hunk 2]
```diff
@@ -416,7 +421,7 @@ public void loadMessageRemoteFailed(Account account, String folder, String uid,
         void onMessageDataLoadFailed();
 
         void onMessageViewInfoLoadFinished(LocalMessage localMessage, MessageViewInfo messageViewInfo);
-        void onMessageViewInfoLoadFailed(LocalMessage localMessage);
+        void onMessageViewInfoLoadFailed(LocalMessage localMessage, MessageViewInfo messageViewInfo);
```
Comment (cketti) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageLoaderHelper.java]: It looks like the `localMessage` argument is never used.

Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageLoaderHelper.java]: I suppose. I only left it for consistency of a message view info load returning callback in `MessageLoaderCallbacks`

Comment (cketti) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageLoaderHelper.java]: The goal is to get rid of having to pass (Local)Message to the view. So being able to remove it here is a win.
Also, there's no need to be consistent with the arguments. We're not writing an API for public consumption. If it's not used, we should remove it.

Comment (Valodim) [inline k9mail/src/main/java/com/fsck/k9/activity/MessageLoaderHelper.java]: now that you mention it, now that #1501 is fixed we can get rid of the LocalMessage parameter in the success case as well :)

**Comment by cketti** [comment]
:+1:

Please rebase and merge.

**Pull request merged**